pandas/core/util/hashing.py

```diff
@@ -255,11 +283,26 @@ def hash_array(
     # hash values. (This check is above the complex check so that we don't ask
     # numpy if categorical is a subdtype of complex, as it will choke).
     if is_categorical_dtype(dtype):
+        vals = cast("Categorical", vals)
         return _hash_categorical(vals, encoding, hash_key)
     elif is_extension_array_dtype(dtype):
         vals, _ = vals._values_for_factorize()
         dtype = vals.dtype
 
+    return _hash_ndarray(vals, encoding, hash_key, categorize)
+
+
+def _hash_ndarray(
```

Comment:
had to separate this out, as i couldn't convince mypy to recognize `vals = cast(np.ndarray, vals)` or `assert isinstance(vals, np.ndarray)`